### PR TITLE
Connections

### PR DESCRIPTION
**Pull request by jgoizueta**
See https://app.clubhouse.io/cartoteam/story/117351 & https://app.clubhouse.io/cartoteam/story/117356

Experimental implementation of connections.

PENDING: (delayed decisions)

* data_imports, synchronizations: add connection_id foreign key, assign from parameters
* consider: connections are not deleted, but deactivated => old imports can be traced to their connection parameters
* full migration support of connections
* encrypt password, token on exports/imports (migrations), hide in logs
* decide: oauth creation pre-creates connection record
* consider: ConnectionManager identifies connections by name => API does as well (impacts previous  decision)

### CODE REVIEW

**Review by dgaubert — APPROVED** [review]
Apart from what Rubocop is claiming, I have a minor question about if we need to define the version of the google's gems in gemfiles or not. Left some comments, nothing blocking.

Note: rubocop's comments make hard to read the code of the PR. Didn't find how to hide them.

LGTM

File: app/controllers/carto/api/public/connections_controller.rb [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        end
+
+        def connect
+          connection = @connection_manager.fetch_connection(params[:id])
+          @connection_manager.check(connection)
+          connector = Carto::Connector.new(parameters: {}, connection: connection, user: current_user, logger: nil)
+          render_jsonp({ connected: @connection_manager.check(connection) }, 200)
+        end
+
+        def projects
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply app/controllers/carto/api/public/connections_controller.rb:91]: Affirmative.

File: app/controllers/carto/api/public/connections_controller.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        end
+
+        private
+
+        def rescue_from_connection_not_found(exception)
+          render_jsonp({ errors: exception.message }, 404)
+        end
+
+        def rescue_from_invalid_connection(exception)
+          code =  exception.message =~ /Access Denied/im ? 401 : 422
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply app/controllers/carto/api/public/connections_controller.rb:109]: You're right!

File: app/controllers/carto/api/public/connections_controller.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            render_jsonp({ id: connection.id }, 201)
+          else
+            # initiate creation of OAuth connection
+            render_jsonp({ auth_url: @connection_manager.create_oauth_connection_get_url(service: params[:connector]) })
+          end
+        end
+
+        def destroy
+          @connection_manager.delete_connection(params[:id])
+          head :ok
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply app/controllers/carto/api/public/connections_controller.rb:45]: yep

File: app/controllers/carto/api/public/connections_controller.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        def update
+          @connection_manager.update_db_connection(id: params[:id], parameters: params[:parameters], name: params[:name])
+          head :ok
+        end
+
+        def check_oauth
+          service = params[:service]
+          connection = @connection_manager.find_oauth_connection(service)
+          # shouldn't it return a presented connection? and raise an exception if not found?
+          if connection
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply app/controllers/carto/api/public/connections_controller.rb:57]: Yes, I think's preferable for clarity/style, even though here connection is either a `Connection` or `nil`, so it shouldn't make any difference (`present?` takes into account other values that are not _falsy_ but are empty like `[]` or blank like `''`)

File: app/controllers/carto/api/public/connections_controller.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+          else
+            render_jsonp({ errors: "Provider #{provider_id} doesn't support dry runs" }, 422)
+          end
+        end
+
+        def connect
+          connection = @connection_manager.fetch_connection(params[:id])
+          @connection_manager.check(connection)
+          connector = Carto::Connector.new(parameters: {}, connection: connection, user: current_user, logger: nil)
+          render_jsonp({ connected: @connection_manager.check(connection) }, 200)
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply app/controllers/carto/api/public/connections_controller.rb:88]: oopsie!

File: lib/carto/connector/connection_manager.rb [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+            available: is_enabled
+          }
+          connector[:connections] = list_connections(connector: provider) if connections
+          connector
+        }
+      end
+
+      # Unify connectors of dual type
+      oauth_connectors.each do |oauth_connector|
+        db_connector = db_connectors.find { |c| c[:connector] == oauth_connector[:connector] }
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply lib/carto/connector/connection_manager.rb:57]: it's initialized to `[]`  at the top, so no problem

**Pull request merged**